File: src/java/us/toadlet/pad/AndroidApplication.java
/********** Copyright header - do not remove **********
 *
 * The Toadlet Engine
 *
 * Copyright 2009, Lightning Toads Productions, LLC
 *
 * Author(s): Alan Fischer, Andrew Fischer
 *
 * This file is part of The Toadlet Engine.
 *
 * The Toadlet Engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * The Toadlet Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public License
 * along with The Toadlet Engine.  If not, see <http://www.gnu.org/licenses/>.
 *
 ********** Copyright header - do not remove **********/

package us.toadlet.pad;

import android.app.Activity;
import android.content.Context;
import android.view.*;
import android.os.*;
import us.toadlet.peeper.*;

class ApplicationView extends SurfaceView implements SurfaceHolder.Callback{
	public ApplicationView(AndroidApplication application){
		super(application);
		mApplication=application;
		requestFocus();
		setFocusableInTouchMode(true);
		getHolder().addCallback(this);
		getHolder().setType(SurfaceHolder.SURFACE_TYPE_GPU);
	}

	public void onSizeChanged(int w,int h,int oldw,int oldh){
		mApplication.notifySizeChanged(w,h);
	}

	public boolean onKeyDown(int keyCode,KeyEvent event){
		mApplication.notifyKeyPressed(event.getUnicodeChar());
		return true;
	}
	
	public boolean onKeyUp(int keyCode,KeyEvent event){
		mApplication.notifyKeyReleased(event.getUnicodeChar());
		return true;
	}

	public boolean onTouchEvent(MotionEvent event){
		int x=(int)event.getX(),y=(int)event.getY();
		
		switch(event.getAction()){
			case MotionEvent.ACTION_DOWN:
				mApplication.notifyMousePressed(x,y);
			break;
			case MotionEvent.ACTION_MOVE:
				mApplication.notifyMouseMoved(x,y);
			break;
			case MotionEvent.ACTION_UP:
				mApplication.notifyMouseReleased(x,y);
			break;
		}

		try{Thread.sleep(100);}catch(Exception ex){}
		return true;
	}

	public void surfaceCreated(SurfaceHolder holder){
		mApplication.notifySurfaceCreated(holder,true);
		mApplication.notifySizeChanged(getWidth(),getHeight());
	}

	public void surfaceDestroyed(SurfaceHolder holder){
		mApplication.notifySurfaceDestroyed(holder,true);
	}

	public void surfaceChanged(SurfaceHolder holder,int format,int width,int height){}

	AndroidApplication mApplication;
}

public abstract class AndroidApplication extends Activity implements RenderTarget,Runnable{
    public AndroidApplication(){
		super();
	}

	protected void onCreate(Bundle bundle){
		super.onCreate(bundle);

		create();
	}

	protected void onStart(){
		System.out.println("AndroidApplication.onStart");
	
		super.onStart();
		start();
	}

	protected void onStop(){
		System.out.println("AndroidApplication.onStop");

		super.onStop();
		stop();
	}

	protected void onDestroy(){
		System.out.println("AndroidApplication.onDestroy");

		super.onStop();
		destroy();
	}

	protected void onPause(){
		super.onPause();
		if(mActive==true){
			notifySurfaceDestroyed(mView.getHolder(),false);
		}
	}

	protected void onResume(){
		super.onResume();
		if(mActive==false){
			notifySurfaceCreated(mView.getHolder(),false);
			notifySizeChanged(mView.getWidth(),mView.getHeight());
		}
	}
	
	public void create(){
//		mEngine=makeEngine();

//		mEngine.installHandlers();
		
//		if(mApplet!=null){
//			mApplet.create();
//		}
	}
	
	public void destroy(){
//		if(mApplet!=null){
//			mApplet.destroy();
//		}

//		deleteEngine(mEngine);
	}
	
	public void start(){
		if(mFullscreen){
			requestWindowFeature(Window.FEATURE_NO_TITLE);   
			getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,WindowManager.LayoutParams.FLAG_FULLSCREEN);
		}
	
		mView=new ApplicationView(this);
		setContentView(mView);

		mLastTime=System.currentTimeMillis();

		mRun=true;
		mThread=new Thread(this);
		mThread.start();
	}
	
	public void stop(){
		finish();
		
		mRun=false;
		
		try{
			mThread.join();
		}catch(InterruptedException ex){}
	}

	protected abstract Applet createApplet(AndroidApplication app);
	protected abstract void destroyApplet(Applet applet);

	public void run(){
		while(mRun){
			long currentTime=System.currentTimeMillis();
			if(mActive){
				update((int)(currentTime-mLastTime));
				if(mRenderDevice!=null){
					render(mRenderDevice);
				}
			}
			mLastTime=currentTime;

			synchronized(mSurfaceMutex){
				if(mNotifySurfaceCreated!=null){
if(mEngine==null){
	mEngine=makeEngine();
}

					surfaceCreated(mNotifySurfaceCreated);

if(mEngine!=null){
	mEngine.installHandlers();
}

if(mApplet==null){
	setApplet(createApplet(this));
	mApplet.create();
}

					mNotifySurfaceCreated=null;
					mActive=true;
					mSurfaceMutex.notify();
				}
				if(mNotifySurfaceDestroyed!=null){
if(mApplet!=null){
	mApplet.destroy();
	destroyApplet(mApplet);
	setApplet(null);
}

if(mEngine!=null){
	mEngine.destroy();
}
					surfaceDestroyed(mNotifySurfaceDestroyed);

if(mEngine!=null){
	deleteEngine(mEngine);
	mEngine=null;
}

					mNotifySurfaceDestroyed=null;
					mActive=false;
					mSurfaceMutex.notify();
				}
				if(mNotifySizeChanged){
					mNotifySizeChanged=false;
					resized(mWidth,mHeight);

					if(mActive && mRenderDevice!=null){
						render(mRenderDevice);
					}
				}
			}
			synchronized(this){
				if(mNotifyKeyPressed){
					mNotifyKeyPressed=false;
					keyPressed(mKeyPressed);
				}
				if(mNotifyKeyReleased){
					mNotifyKeyReleased=false;
					keyReleased(mKeyReleased);
				}
				if(mNotifyMousePressed){
					mNotifyMousePressed=false;
					mousePressed(mMousePressedX,mMousePressedY,0);
				}
				if(mNotifyMouseMoved){
					mNotifyMouseMoved=false;
					mouseMoved(mMouseMovedX,mMouseMovedY);
				}
				if(mNotifyMouseReleased){
					mNotifyMouseReleased=false;
					if(mMouseMovedX!=mMouseReleasedX || mMouseMovedY!=mMouseReleasedY){
						mouseMoved(mMouseReleasedX,mMouseReleasedY);
					}
					mouseReleased(mMouseReleasedX,mMouseReleasedY,0);
				}
			}
		}
	}

	synchronized void notifyKeyPressed(int key){
		if(mNotifyKeyPressed==false){
			mNotifyKeyPressed=true;
			mKeyPressed=key;
		}
	}

	synchronized void notifyKeyReleased(int key){
		if(mNotifyKeyReleased==false){
			mNotifyKeyReleased=true;
			mKeyReleased=key;
		}
	}

	synchronized void notifyMousePressed(int x,int y){
		if(mNotifyMousePressed==false){
			mNotifyMousePressed=true;
			mMousePressedX=x;
			mMousePressedY=y;
		}
	}

	synchronized void notifyMouseMoved(int x,int y){
		if(mNotifyMouseMoved==false){
			mNotifyMouseMoved=true;
			mMouseMovedX=x;
			mMouseMovedY=y;
		}
	}

	synchronized void notifyMouseReleased(int x,int y){
		if(mNotifyMouseReleased==false){
			mNotifyMouseReleased=true;
			mMouseReleasedX=x;
			mMouseReleasedY=y;
		}
	}

	void notifySizeChanged(int width,int height){
		synchronized(mSurfaceMutex){
			mNotifySizeChanged=true;
			mWidth=width;
			mHeight=height;
		}
	}
	
	void notifySurfaceCreated(SurfaceHolder holder,boolean fromSurface){
		synchronized(mSurfaceMutex){
			if(mRun){
				if(fromSurface==true || mSurfaceCreated){
					mNotifySurfaceCreated=holder;
					try{ mSurfaceMutex.wait(); }
					catch(Exception ex){}
				}
			}
			else{
				surfaceDestroyed(holder);
			}

			if(fromSurface){
				mSurfaceCreated=true;
			}
		}
	}

	void notifySurfaceDestroyed(SurfaceHolder holder,boolean fromSurface){
		synchronized(mSurfaceMutex){
			if(mRun){
				if(fromSurface==true || mSurfaceCreated){
					mNotifySurfaceDestroyed=holder;
					try{ mSurfaceMutex.wait(); }
					catch(Exception ex){}
				}
			}
			else{
				surfaceDestroyed(holder);
			}

			if(fromSurface){
				mSurfaceCreated=false;
			}
		}
	}

	public boolean isRunning(){return mRun;}

	public void setAutoActivate(boolean autoActivate){}
	public boolean getAutoActivate(){return false;}
	
	public boolean isActive(){return mActive;}

	public void setPosition(int x,int y){}
	public int getPositionX(){return 0;}
	public int getPositionY(){return 0;}

	public void setSize(int width,int height){}
	public int getWidth(){return mView==null?0:mView.getWidth();}
	public int getHeight(){return mView==null?0:mView.getHeight();}

	public void setFullscreen(boolean fullscreen){mFullscreen=fullscreen;}
	public boolean getFullscreen(){return mFullscreen;}
	
	public void setApplet(Applet applet){mApplet=applet;}
	public Applet getApplet(){return mApplet;}

	public RenderTarget getRootRenderTarget(){return mRenderTarget;}
	public boolean isPrimary(){return mRenderTarget.isPrimary();}
	public boolean isValid(){return mRenderTarget.isValid();}

	public Engine getEngine(){return mEngine;}
	public RenderDevice getRenderDevice(){return mRenderDevice;}

	public void resized(int width,int height)		{if(mApplet!=null)mApplet.resized(width,height);}
	public void focusGained()						{if(mApplet!=null)mApplet.focusGained();}
	public void focusLost()							{if(mApplet!=null)mApplet.focusLost();}
	public void keyPressed(int key)					{if(mApplet!=null)mApplet.keyPressed(key);}
	public void keyReleased(int key)				{if(mApplet!=null)mApplet.keyReleased(key);}
	public void mouseMoved(int x,int y)				{if(mApplet!=null)mApplet.mouseMoved(x,y);}
	public void mousePressed(int x,int y,int button){if(mApplet!=null)mApplet.mousePressed(x,y,button);}
	public void mouseReleased(int x,int y,int button){if(mApplet!=null)mApplet.mouseReleased(x,y,button);}
	public void mouseScrolled(int x,int y,int scroll){if(mApplet!=null)mApplet.mouseScrolled(x,y,scroll);}
	public void update(int dt)						{if(mApplet!=null)mApplet.update(dt);}
	public void render(RenderDevice renderDevice)	{if(mApplet!=null)mApplet.render(renderDevice);}

	public void surfaceCreated(SurfaceHolder holder){
		System.out.println(
			"AndroidApplication.surfaceCreated");

		boolean result=false;
		mRenderTarget=makeRenderTarget(holder);
		if(mRenderTarget!=null){
			mRenderDevice=makeRenderDevice();

			result=mRenderDevice.create(this,null);
			if(result==false){
				mRenderDevice=null;
			}
		}
		if(result==false){
			mRenderTarget=null;

			System.out.println(
				"Error creating RenderTargetPeer");
		}
		else if(mRenderDevice==null){
			System.out.println(
				"error creating RenderDevice");
		}

		if(mRenderDevice!=null){
			mRenderDevice.setRenderTarget(this);
			mEngine.setRenderDevice(mRenderDevice);
		}
	}

	public void surfaceDestroyed(SurfaceHolder holder){
		System.out.println(
			"AndroidApplication.surfaceDestroyed");

		if(mRenderDevice!=null){
			mEngine.setRenderDevice(null);
			mRenderDevice.destroy();
			deleteRenderDevice(mRenderDevice);
			mRenderDevice=null;
		}

		if(mRenderTarget!=null){
			mRenderTarget=null;
		}
	}

	protected RenderTarget makeRenderTarget(SurfaceHolder holder){
		WindowRenderTargetFormat format=new WindowRenderTargetFormat();
		format.pixelFormat=TextureFormat.Format_RGB_5_6_5;
		format.depthBits=16;
		format.multisamples=0;
		format.threads=0;
		RenderTarget target=new EGLWindowRenderTarget(holder,format);
		if(target!=null && target.isValid()==false){
			target=null;
		}
		return target;
	}

	protected native Engine makeEngine();
	protected native void deleteEngine(Engine engine);
	protected native RenderDevice makeRenderDevice();
	protected native void deleteRenderDevice(RenderDevice device);
	
	Thread mThread=null;
	protected ApplicationView mView;
	protected Applet mApplet;
	protected Engine mEngine;
	protected boolean mFullscreen;
	protected boolean mActive;
	protected boolean mRun;
	protected Object mSurfaceMutex=new Object();
	protected long mLastTime=0;
	protected RenderTarget mRenderTarget;
	protected RenderDevice mRenderDevice;
//	protected AudioDevice mAudioDevice;
//	protected MotionDevice mMotionDevice;
		
	protected SurfaceHolder mNotifySurfaceCreated;
	protected SurfaceHolder mNotifySurfaceDestroyed;
	protected boolean mSurfaceCreated;
	protected boolean mNotifySizeChanged;
	protected int mWidth,mHeight;
	protected boolean mNotifyKeyPressed;
	protected int mKeyPressed;
	protected boolean mNotifyKeyReleased;
	protected int mKeyReleased;
	protected boolean mNotifyMousePressed;
	protected int mMousePressedX,mMousePressedY;
	protected boolean mNotifyMouseMoved;
	protected int mMouseMovedX,mMouseMovedY;
	protected boolean mNotifyMouseReleased;
	protected int mMouseReleasedX,mMouseReleasedY;
}

File: source/cpp/toadlet/pad/platform/android/AndroidApplication.cpp
/********** Copyright header - do not remove **********
 *
 * The Toadlet Engine
 *
 * Copyright 2009, Lightning Toads Productions, LLC
 *
 * Author(s): Alan Fischer, Andrew Fischer
 *
 * This file is part of The Toadlet Engine.
 *
 * The Toadlet Engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * The Toadlet Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public License
 * along with The Toadlet Engine.  If not, see <http://www.gnu.org/licenses/>.
 *
 ********** Copyright header - do not remove **********/

#include "AndroidApplication.h"
#include <toadlet/egg/System.h>
#include <toadlet/egg/platform/android/AndroidAssetArchive.h>
#include <toadlet/tadpole/platform/android/AndroidTextureStreamer.h>
#include <android/sensor.h>

using namespace toadlet::egg;
using namespace toadlet::peeper;
using namespace toadlet::ribbit;
using namespace toadlet::flick;
using namespace toadlet::tadpole;
using namespace toadlet::pad;

TOADLET_C_API RenderTarget *new_EGLWindowRenderTarget(void *display,void *window,WindowRenderTargetFormat *format);
TOADLET_C_API RenderDevice *new_GLES1RenderDevice();
TOADLET_C_API RenderDevice *new_GLES2RenderDevice();
TOADLET_C_API InputDevice *new_AndroidSensorDevice(int type);

namespace toadlet{
namespace pad{

AndroidApplication::AndroidApplication():
	mActivity(NULL),
	mConfig(NULL),
	mLooper(NULL),
	mQueue(NULL),
	mWindow(NULL),mNotifyWindowCreated(NULL),mNotifyWindowDestroyed(NULL),
	mNotifyWindowResized(false),
	mNotifyQueueCreated(NULL),mNotifyQueueDestroyed(NULL),
	mWidth(0),mHeight(0),
	mLastX(0),mLastY(0),
	mDifferenceMouse(false),
	//mThread,
	mRun(false),

	mEngine(NULL),
	mRenderDevice(NULL),
	mAudioDevice(NULL),
	mApplet(NULL)
{
	mFormat=WindowRenderTargetFormat::ptr(new WindowRenderTargetFormat());
	mFormat->setDepthBits(16);
	mFormat->setFlags(2);
	#if defined(TOADLET_DEBUG)
		mFormat->setDebug(true);
	#else
		mFormat->setDebug(false);
	#endif
}

AndroidApplication::~AndroidApplication(){
}

bool AndroidApplication::create(String renderDevice,String audioDevice){
	mConfig=AConfiguration_new();

	JNIEnv *env=mActivity->env;
	jobject obj=mActivity->clazz;

	mEngine=new Engine(env,obj);
	
	mEngine->installHandlers();

	AndroidTextureStreamer::ptr textureStreamer=new AndroidTextureStreamer(mEngine->getTextureManager(),env);
	mEngine->getTextureManager()->setDefaultStreamer(textureStreamer);
	
	InputDevice::ptr linearDevice=new_AndroidSensorDevice(ASENSOR_TYPE_ACCELEROMETER);
	if(linearDevice->create()){
		mInputDevices[linearDevice->getType()]=linearDevice;
	}

	InputDevice::ptr angularDevice=new_AndroidSensorDevice(ASENSOR_TYPE_GYROSCOPE);
	if(angularDevice->create()){
		mInputDevices[angularDevice->getType()]=angularDevice;
	}

	InputDevice::ptr lightDevice=new_AndroidSensorDevice(ASENSOR_TYPE_LIGHT);
	if(lightDevice->create()){
		mInputDevices[lightDevice->getType()]=lightDevice;
	}

	InputDevice::ptr proximityDevice=new_AndroidSensorDevice(ASENSOR_TYPE_PROXIMITY);
	if(proximityDevice->create()){
		mInputDevices[proximityDevice->getType()]=proximityDevice;
	}

	InputDevice::ptr magneticDevice=new_AndroidSensorDevice(ASENSOR_TYPE_MAGNETIC_FIELD);
	if(magneticDevice->create()){
		mInputDevices[magneticDevice->getType()]=magneticDevice;
	}

	if(mApplet!=NULL){
		mApplet->create();
	}

	activate();
}

void AndroidApplication::destroy(){
	if(mEngine!=NULL){
		mEngine->destroy();
	}

	deactivate();

	if(mApplet!=NULL){
		mApplet->destroy();
		mApplet=NULL;
	}

	mEngine=NULL;

	int i;
	for(i=0;i<InputDevice::InputType_MAX;++i){
		if(mInputDevices[i]!=NULL){
			mInputDevices[i]->destroy();
			mInputDevices[i]=NULL;
		}
	}

	if(mQueue!=NULL){
        AInputQueue_detachLooper(mQueue);
		mQueue=NULL;
	}
	
	if(mConfig!=NULL){
		AConfiguration_delete(mConfig);
		mConfig=NULL;
	}
	
	mLooper=NULL;
}

void AndroidApplication::start(){
	mRun=true;
	mThread=Thread::ptr(new Thread(this));
	mThread->start();
}

void AndroidApplication::stop(){
	mRun=false;
	// pthread_join on android currently hangs here
//	mThread->join();
	mThread=NULL;
}

bool AndroidApplication::isRunning() const{
	return mRun;
}

void AndroidApplication::notifyWindowCreated(ANativeWindow *window){
	Log::alert("notifyWindowCreated");

	mWindowMutex.lock();
		if(mRun){
			mNotifyWindowCreated=window;
			mWindowCondition.wait(&mWindowMutex);
		}
		else{
			windowCreated(window);
		}
	mWindowMutex.unlock();
}

void AndroidApplication::notifyWindowDestroyed(ANativeWindow *window){
	Log::alert("notifyWindowDestroyed");

	mWindowMutex.lock();
		if(mRun){
			mNotifyWindowDestroyed=window;
			mWindowCondition.wait(&mWindowMutex);
		}
		else{
			windowDestroyed(window);
		}
	mWindowMutex.unlock();
}

void AndroidApplication::notifyWindowResized(){
	Log::alert("onNativeWindowResized");

	mWindowMutex.lock();
		if(mRun){
			mNotifyWindowResized=true;
			mWindowCondition.wait(&mWindowMutex);
		}
		else{
			windowResized();
		}
	mWindowMutex.unlock();
}

void AndroidApplication::notifyQueueCreated(AInputQueue *queue){
	Log::alert("notifyQueueCreated");

	mWindowMutex.lock();
		if(mRun){
			mNotifyQueueCreated=queue;
			mWindowCondition.wait(&mWindowMutex);
		}
		else{
			queueCreated(queue);
		}
	mWindowMutex.unlock();
}

void AndroidApplication::notifyQueueDestroyed(AInputQueue *queue){
	Log::alert("notifyQueueDestroyed");

	mWindowMutex.lock();
		if(mRun){
			mNotifyQueueDestroyed=queue;
			mWindowCondition.wait(&mWindowMutex);
		}
		else{
			queueDestroyed(queue);
		}
	mWindowMutex.unlock();
}

void AndroidApplication::setNativeActivity(ANativeActivity *activity){
	activity->instance=this;

	activity->callbacks->onDestroy = onDestroy;
	activity->callbacks->onStart = onStart;
//	activity->callbacks->onResume = onResume;
//	activity->callbacks->onSaveInstanceState = onSaveInstanceState;
//	activity->callbacks->onPause = onPause;
	activity->callbacks->onStop = onStop;
	activity->callbacks->onConfigurationChanged = onConfigurationChanged;
//	activity->callbacks->onLowMemory = onLowMemory;
//	activity->callbacks->onWindowFocusChanged = onWindowFocusChanged;
	activity->callbacks->onNativeWindowCreated = onNativeWindowCreated;
	activity->callbacks->onNativeWindowDestroyed = onNativeWindowDestroyed;
	activity->callbacks->onNativeWindowResized = onNativeWindowResized;
	activity->callbacks->onInputQueueCreated = onInputQueueCreated;
	activity->callbacks->onInputQueueDestroyed = onInputQueueDestroyed;

	mActivity=activity;

	create("","");
}

void AndroidApplication::run(){
	mLooper=ALooper_prepare(ALOOPER_PREPARE_ALLOW_NON_CALLBACKS);
	if(mQueue!=NULL){
		AInputQueue_attachLooper(mQueue,mLooper,0,NULL,this);
	}
	
	uint64 lastTime=System::mtime();
	while(mRun){
		stepEventLoop();
	
		uint64 currentTime=System::mtime();
		int dt=currentTime-lastTime;

		update(dt);
		
		if(mRenderDevice!=NULL){
			// onNativeWindowResized doesn't appear to be called properly, so instead we check each frame for a resize
			if(mRenderTarget->getWidth()!=mWidth || mRenderTarget->getHeight()!=mHeight){
				windowResized();
			}
		
			if(mRenderDevice->activate()==RenderDevice::DeviceStatus_NEEDSRESET){
				mEngine->contextReset(mRenderDevice);
			}

			render();
		}

		if(mAudioDevice!=NULL){
			mAudioDevice->update(dt);
		}

		mWindowMutex.lock();
			if(mNotifyWindowCreated!=NULL){
				windowCreated(mNotifyWindowCreated);
				
				mNotifyWindowCreated=NULL;
				mWindowCondition.notify();
			}
			if(mNotifyWindowDestroyed!=NULL){
				windowDestroyed(mNotifyWindowDestroyed);
				
				mNotifyWindowDestroyed=NULL;
				mWindowCondition.notify();
			}
			if(mNotifyWindowResized){
				windowResized();
				
				mNotifyWindowResized=false;
				mWindowCondition.notify();
			}
			if(mNotifyQueueCreated!=NULL){
				queueCreated(mNotifyQueueCreated);
				
				mNotifyQueueCreated=NULL;
				mWindowCondition.notify();
			}
			if(mNotifyQueueDestroyed!=NULL){
				queueDestroyed(mNotifyQueueDestroyed);

				mNotifyQueueDestroyed=NULL;
				mWindowCondition.notify();
			}
		mWindowMutex.unlock();

		lastTime=currentTime;

		System::msleep(0);
	}

	if(mQueue!=NULL){
		AInputQueue_detachLooper(mQueue);
	}
	mLooper=NULL;
}

void AndroidApplication::stepEventLoop(){
	if(mQueue==NULL){
		return;
	}

	int events=0;
	while(ALooper_pollAll(0,NULL,&events,NULL)>=0){
		AInputEvent *event=NULL;
		if(AInputQueue_getEvent(mQueue,&event)>=0){
			if(AInputQueue_preDispatchEvent(mQueue,event)){
				return;
			}
			int32_t handled=0;

			switch(AInputEvent_getType(event)){
				case AINPUT_EVENT_TYPE_KEY:{
					int key=translateKey(AKeyEvent_getKeyCode(event));
					int action=AKeyEvent_getAction(event);
					if(action==AKEY_EVENT_ACTION_DOWN){
						keyPressed(key);
					}
					else if(action==AKEY_EVENT_ACTION_UP){
						keyReleased(key);
					}
				} break;
				case AINPUT_EVENT_TYPE_MOTION:{
					int x=AMotionEvent_getX(event,0),y=AMotionEvent_getY(event,0);
					int action=AMotionEvent_getAction(event);
					if(action==AMOTION_EVENT_ACTION_DOWN){
						mousePressed(x,y,0);
					}
					else if(action==AMOTION_EVENT_ACTION_MOVE){
						if(mDifferenceMouse){
							mouseMoved(mLastX-x,mLastY-y);
						}
						else{
							mouseMoved(x,y);
						}
					}
					else if(action==AMOTION_EVENT_ACTION_UP || action==AMOTION_EVENT_ACTION_CANCEL){
						if(mLastX!=x || mLastY!=y){
							if(mDifferenceMouse){
								mouseMoved(mLastX-x,mLastY-y);
							}
							else{
								mouseMoved(x,y);
							}
						}
						mouseReleased(x,y,0);
					}
					mLastX=x;mLastY=y;
				} break;
			}
			
			AInputQueue_finishEvent(mQueue,event,handled);
		}
	}
}

void AndroidApplication::windowCreated(ANativeWindow *window){
	mWindow=window;

	int nativeFormat=ANativeWindow_getFormat(window);
	if(nativeFormat==WINDOW_FORMAT_RGB_565){
		mFormat->setPixelFormat(TextureFormat::Format_RGB_5_6_5);
	}
	else{
		mFormat->setPixelFormat(TextureFormat::Format_RGBA_8);
	}

	// Start with gles2 and then try gles
	for(;mFormat->getFlags()>=0 && mRenderDevice==NULL;mFormat->setFlags(mFormat->getFlags()-2)){
		RenderTarget::ptr target;
		TOADLET_TRY
			target=new_EGLWindowRenderTarget(0,mWindow,mFormat);
		TOADLET_CATCH_ANONYMOUS(){target=NULL;}
		if(target!=NULL && target->isValid()==false){
			target->destroy();
			target=NULL;
		}

		RenderDevice::ptr device;
		if(target!=NULL){
			TOADLET_TRY
				if(mFormat->getFlags()==2){
					device=new_GLES2RenderDevice();
				}
				else{
					device=new_GLES1RenderDevice();
				}
			TOADLET_CATCH_ANONYMOUS(){device=NULL;}
			if(device!=NULL && device->create(target,0)==false){
				device->destroy();
				target->destroy();
				device=NULL;
				target=NULL;
			}
		}
		
		mRenderTarget=target;
		mRenderDevice=device;
	}
	
	if(mRenderDevice!=NULL){
		mEngine->setRenderDevice(mRenderDevice);
	}
}

void AndroidApplication::windowDestroyed(ANativeWindow *window){
	if(mRenderDevice!=NULL){
		mEngine->setRenderDevice(NULL);
		mRenderDevice->destroy();
		mRenderDevice=NULL;
	}

	if(mRenderTarget!=NULL){
		mRenderTarget->destroy();
		mRenderTarget=NULL;
	}

	mWindow=NULL;
}

void AndroidApplication::windowResized(){
	if(mRenderTarget!=NULL && mRenderDevice!=NULL){
		mWidth=mRenderTarget->getWidth();
		mHeight=mRenderTarget->getHeight();
		resized(mWidth,mHeight);

		render();
	}
}

void AndroidApplication::queueCreated(AInputQueue *queue){
	mQueue=queue;
	AInputQueue_attachLooper(mQueue,mLooper,0,NULL,this);
}

void AndroidApplication::queueDestroyed(AInputQueue *queue){
	AInputQueue_detachLooper(mQueue);
	mQueue=NULL;
}

int AndroidApplication::translateKey(int key){
	switch(key){
		case AKEYCODE_SOFT_LEFT:
			return Key_SOFTLEFT;
		case AKEYCODE_SOFT_RIGHT:
			return Key_SOFTRIGHT;
		case AKEYCODE_BACK:
			return Key_BACK;
		case AKEYCODE_STAR:
			return '*';
		case AKEYCODE_POUND:
			return '#';
		case AKEYCODE_DPAD_UP:
			return Key_UP;
		case AKEYCODE_DPAD_DOWN:
			return Key_DOWN;
		case AKEYCODE_DPAD_LEFT:
			return Key_LEFT;
		case AKEYCODE_DPAD_RIGHT:
			return Key_RIGHT;
		case AKEYCODE_DPAD_CENTER:
			return Key_ACTION;
		case AKEYCODE_COMMA:
			return ',';
		case AKEYCODE_PERIOD:
			return '.';
		case AKEYCODE_TAB:
			return Key_TAB;
		case AKEYCODE_SPACE:
			return Key_SPACE;
		case AKEYCODE_ENTER:
			return Key_ENTER;
		case AKEYCODE_SHIFT_LEFT:
		case AKEYCODE_SHIFT_RIGHT:
			return Key_SHIFT;
		case AKEYCODE_ALT_LEFT:
		case AKEYCODE_ALT_RIGHT:
			return Key_ALT;
		case AKEYCODE_DEL:
			return Key_DELETE;
		case AKEYCODE_SYM:
			return Key_SPECIAL;
	}

	if(key>=AKEYCODE_0 && key<=AKEYCODE_9){
		return key-AKEYCODE_0+'0';
	}
	else if(key>=AKEYCODE_A && key<=AKEYCODE_Z){
		return key-AKEYCODE_A+'A';
	}

	return key;
}

void AndroidApplication::onDestroy(ANativeActivity *activity){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->destroy();
}

void AndroidApplication::onStart(ANativeActivity *activity){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->start();
}

void AndroidApplication::onStop(ANativeActivity *activity){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->stop();
}

void AndroidApplication::onNativeWindowCreated(ANativeActivity *activity,ANativeWindow *window){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->notifyWindowCreated(window);
}

void AndroidApplication::onNativeWindowDestroyed(ANativeActivity *activity,ANativeWindow *window){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->notifyWindowDestroyed(window);
}

void AndroidApplication::onNativeWindowResized(ANativeActivity *activity,ANativeWindow *window){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->notifyWindowResized();
}

void AndroidApplication::onInputQueueCreated(ANativeActivity *activity,AInputQueue *queue){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->notifyQueueCreated(queue);
}

void AndroidApplication::onInputQueueDestroyed(ANativeActivity *activity,AInputQueue *queue){
	AndroidApplication *app=(AndroidApplication*)activity->instance;
	app->notifyQueueDestroyed(queue);
}

void AndroidApplication::onConfigurationChanged(ANativeActivity *activity){
}

}
}
